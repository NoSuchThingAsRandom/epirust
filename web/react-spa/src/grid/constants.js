const BaseColors = {
    YELLOW: "#fcce44",
    PURPLE: "#3498db",
    BLUE: "#9b59b6",
    RED: "#e74c3c"
}

export const AreaColors = {
    HOUSING: BaseColors.YELLOW,
    WORK: BaseColors.PURPLE,
    TRANSPORT: BaseColors.BLUE,
    HOSPITAL: BaseColors.RED
}

