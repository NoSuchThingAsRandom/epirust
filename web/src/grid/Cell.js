import React from "react";
import PropTypes from 'prop-types'
import {Color} from "./Color";
import "./Cell.scss"

export default function Cell({cellId, color, size}) {
    return (
        <div className={`grid-cell-${cellId}`} style={{width: `${size}px`, height: `${size}px`}}>
            <div className="cell-body"
                 style={{background: `rgb(${color.r}, ${color.g}, ${color.b})`}}/>
        </div>
    );
}

Cell.propTypes = {
    cellId: PropTypes.number.isRequired,
    color: PropTypes.instanceOf(Color).isRequired,
    size: PropTypes.number.isRequired
};