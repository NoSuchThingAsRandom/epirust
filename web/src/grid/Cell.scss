div[class^="grid-cell"] {
  padding: 0 1px 1px 0;
  display: table-cell;
}

.cell-body {
  vertical-align: middle;
  height: 100%;
  width: 100%;
}
