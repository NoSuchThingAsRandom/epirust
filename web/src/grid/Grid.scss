div[class^="grid-row"] {
  display: table-row;
}
