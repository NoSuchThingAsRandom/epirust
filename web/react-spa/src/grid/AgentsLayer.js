/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import React, { useContext, useRef, useState, useEffect } from 'react';
import { GridContext } from './index'
import { BaseColors } from './constants';


export default function AgentPositionsWrapper({ agentPositions }) {
    const [agentPositionsPerHour, setAgentPositionsPerHour] = useState([])

    useEffect(() => {

        agentPositions.forEach(({ citizen_states }, index) => {
            setTimeout(() => {
                setAgentPositionsPerHour(citizen_states)
            }, (index + 1) * 500)
        })

    }, [agentPositions])

    return (<AgentsLayer agentPositionsPerHour={agentPositionsPerHour} />)
}


function AgentsLayer({ agentPositionsPerHour }) {
    const { cellDimension, lineWidth, canvasDimension } = useContext(GridContext);

    const agentsLayerCanvas = useRef(null);
    const [agentsCanvasContext, setAgentsCanvasContext] = useState(null);

    useEffect(() => {
        if (!agentsLayerCanvas)
            return

        setAgentsCanvasContext(agentsLayerCanvas.current.getContext("2d"));

    }, [agentsLayerCanvas])

    useEffect(() => {
        if (!agentsCanvasContext)
            return

        agentsCanvasContext.clearRect(0, 0, canvasDimension, canvasDimension);
        agentsCanvasContext.fillStyle = BaseColors.BLACK
        agentPositionsPerHour.forEach((agent) => {
            const { x, y } = agent.location
            agentsCanvasContext.beginPath();
            const agentCircleRadius = Math.floor(cellDimension / 4)
            const startAngle = 0, endAngle = 2 * Math.PI
            agentsCanvasContext.arc(calculateCoordinate(x), calculateCoordinate(y), agentCircleRadius, startAngle, endAngle);
            agentsCanvasContext.fill();
        })
    }, [agentsCanvasContext, agentPositionsPerHour])

    function calculateCoordinate(x) {
        return Math.floor(x * cellDimension + (0.5 * cellDimension) + lineWidth)
    }

    return (
        <canvas ref={agentsLayerCanvas} id="grid-canvas-agents" width={canvasDimension} height={canvasDimension} style={{ position: "absolute", zIndex: 4 }} />
    )
}
