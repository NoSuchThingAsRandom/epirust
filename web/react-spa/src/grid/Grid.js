/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import PropTypes from 'prop-types'
import Cell from "./Cell";
import {Color} from "./Color";
import "./Grid.scss"

export default function Grid({size}) {
    let color = new Color(75, 34, 23);
    let cellSize = 10;

    function repeat(element, times) {
        let _range = [];
        for (let i = 0; i < times; i++) {
            _range.push(element(i));
        }
        return _range;
    }

    const cell = (cellId) => <Cell key={cellId} cellId={cellId} size={cellSize} color={color}/>;
    const row = (rowId) => (
        <div className={`grid-row-${rowId}`} key={rowId}>
            {repeat(cell, size).map(x => x)}
        </div>
    );

    return (
        <div className="grid">
            <div className="row">
                <div className="col-8">
                    {repeat(row, size).map(x => x)}
                </div>
                <div className="col-4">
                    Config
                </div>
            </div>
        </div>
    );
}

Grid.propTypes = {
    size: PropTypes.number.isRequired
};
