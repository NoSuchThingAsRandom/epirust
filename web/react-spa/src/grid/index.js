/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from 'react';

import gridLayout from '../resources/grid-layout';
import LandmarksLayer from './LandmarksLayer';
import LinesLayer from './LinesLayer';
import AreasLayer from './AreasLayer';
import AgentsLayer from './AgentsLayer';
import { AreaColors, AgentStateToColor } from './constants';
import agentMovement from '../resources/agent-movement';


export default function GridPage() {
    const { housing_area, work_area, transport_area, hospital_area } = gridLayout
    const areaDimensions = [
        { ...housing_area, color: AreaColors.HOUSING },
        { ...work_area, color: AreaColors.WORK },
        { ...transport_area, color: AreaColors.TRANSPORT },
        { ...hospital_area, color: AreaColors.HOSPITAL }
    ]

    const housesDimensions = gridLayout.houses,
        officesDimensions = gridLayout.offices

    return (
        <div className="grid-wrap">
            <CanvasGrid size={gridLayout.grid_size} areaDimensions={areaDimensions} landmarksDimensions={{ housesDimensions, officesDimensions }} />
            <GridLegends />
        </div>
    )
}

function GridLegends() {

    const areasLegends = [
        { backgroundColor: AreaColors.HOUSING, text: "Housing" },
        { backgroundColor: AreaColors.WORK, text: "Work" },
        { backgroundColor: AreaColors.TRANSPORT, text: "Transport" },
        { backgroundColor: AreaColors.HOSPITAL, text: "Hospital" },
        { backgroundColor: AreaColors.OTHER, text: "Other" }
    ]

    const agentsLegends = [
        { backgroundColor: AgentStateToColor.s, text: "Susceptible" },
        { backgroundColor: AgentStateToColor.i, text: "Infected" },
        { backgroundColor: AgentStateToColor.r, text: "Recovered" },
        { backgroundColor: AgentStateToColor.d, text: "Deceased" }
    ]
    return (
        <div className="legends-wrap">
            <h6>Areas</h6>
            <ul className="legends areas">{
                areasLegends.map(({ backgroundColor, text }) => (
                    <li>
                        <div className="legend-item">
                            <span className="swatch" style={{ backgroundColor }}></span>
                            <span className="text">{text}</span>
                        </div>
                    </li>
                ))}
            </ul>

            <h6>Agent States</h6>
            <ul className="legends agents">{
                agentsLegends.map(({ backgroundColor, text }) => (
                    <li>
                        <div className="legend-item">
                            <span className="swatch" style={{ backgroundColor }}></span>
                            <span className="text">{text}</span>
                        </div>
                    </li>
                ))}
            </ul>
        </div>
    )
}

export const GridContext = React.createContext(null);

export function CanvasGrid({ size, areaDimensions, landmarksDimensions }) {

    const cellDimension = Math.floor((window.innerHeight - 165) / size),
        lineWidth = Math.floor(cellDimension / 4) < 1 ? 0 : Math.floor(cellDimension / 4),
        canvasDimension = (size * cellDimension) + lineWidth;

    return (
        <div style={{ position: "relative" }}>
            <GridContext.Provider value={{
                cellDimension: cellDimension,
                lineWidth: lineWidth,
                canvasDimension: canvasDimension,
                size: size
            }}>
                <AreasLayer areaDimensions={areaDimensions} />
                <LinesLayer />
                <LandmarksLayer landmarksDimensions={landmarksDimensions} />
                <AgentsLayer agentPositions={agentMovement.agentsPerTick} />
            </GridContext.Provider>
        </div>
    )
}