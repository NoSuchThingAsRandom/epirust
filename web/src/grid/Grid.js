import React from "react";
import PropTypes from 'prop-types'
import Cell from "./Cell";
import {Color} from "./Color";
import "./Grid.scss"

export default function Grid({size}) {
    let color = new Color(75, 34, 23);
    let cellSize = 10;

    function repeat(element, times) {
        let _range = [];
        for (let i = 0; i < times; i++) {
            _range.push(element(i));
        }
        return _range;
    }

    const cell = (cellId) => <Cell key={cellId} cellId={cellId} size={cellSize} color={color}/>;
    const row = (rowId) => (
        <div className={`grid-row-${rowId}`} key={rowId}>
            {repeat(cell, size).map(x => x)}
        </div>
    );

    return (
        <div className="grid">
            <div className="row">
                <div className="col-8">
                    {repeat(row, size).map(x => x)}
                </div>
                <div className="col-4">
                    Config
                </div>
            </div>
        </div>
    );
}

Grid.propTypes = {
    size: PropTypes.number.isRequired
};