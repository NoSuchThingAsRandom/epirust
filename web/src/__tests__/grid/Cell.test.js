/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import renderer from 'react-test-renderer'
import Cell from "../../grid/Cell";
import {Color} from "../../grid/Color";
import React from "react";


describe('Cell', function () {
    it('should render with specified color and size', function () {
        const testColor = new Color(10, 11, 12);
        const component = renderer.create(<Cell color={testColor} size={10} cellId={1}/>);

        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });
});
