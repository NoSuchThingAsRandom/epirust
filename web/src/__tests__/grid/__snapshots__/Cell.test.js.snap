// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should render with specified color and size 1`] = `
<div
  className="grid-cell-1"
  style={
    Object {
      "height": "10px",
      "width": "10px",
    }
  }
>
  <div
    className="cell-body"
    style={
      Object {
        "background": "rgb(10, 11, 12)",
      }
    }
  />
</div>
`;
